File: sdk/session/session.go
package s4wave_session

import (
	"context"

	resource_client "github.com/s4wave/spacewave/bldr/resource/client"
	session "github.com/s4wave/spacewave/core/session"
)

// Session is a session resource that provides access to session functionality.
// The MountSession directive will remain active until this resource is released.
//
// This Go SDK implementation wraps SessionResourceService.
type Session struct {
	// client supplies the borrowed resource connection.
	client *resource_client.Client
	// ref retains the mounted Session until Release.
	ref resource_client.ResourceRef
	// service invokes the mounted Session resource.
	service SRPCSessionResourceServiceClient
}

// NewSession creates a new Session resource wrapper.
func NewSession(client *resource_client.Client, ref resource_client.ResourceRef) (*Session, error) {
	srpcClient, err := ref.GetClient()
	if err != nil {
		return nil, err
	}
	return &Session{
		client:  client,
		ref:     ref,
		service: NewSRPCSessionResourceServiceClient(srpcClient),
	}, nil
}

// GetResourceRef returns the resource reference.
func (s *Session) GetResourceRef() resource_client.ResourceRef {
	return s.ref
}

// Release releases the resource reference.
func (s *Session) Release() {
	s.ref.Release()
}

// GetSessionInfo returns information about this session.
func (s *Session) GetSessionInfo(ctx context.Context) (*GetSessionInfoResponse, error) {
	return s.service.GetSessionInfo(ctx, &GetSessionInfoRequest{})
}

// CreateSpace creates a new Space as a SharedObject within the Session.
// ownerType is "account" or "organization"; ownerID is the principal id
// (account id or organization id).
func (s *Session) CreateSpace(ctx context.Context, spaceName, ownerType, ownerID string) (*CreateSpaceResponse, error) {
	return s.service.CreateSpace(ctx, &CreateSpaceRequest{
		SpaceName: spaceName,
		OwnerType: ownerType,
		OwnerId:   ownerID,
	})
}

// DeleteSpace deletes a space and its associated storage.
func (s *Session) DeleteSpace(ctx context.Context, sharedObjectID string) (*DeleteSpaceResponse, error) {
	return s.service.DeleteSpace(ctx, &DeleteSpaceRequest{SharedObjectId: sharedObjectID})
}

// LeaveSpace relinquishes native participation while retaining existing local data.
func (s *Session) LeaveSpace(ctx context.Context, sharedObjectID string) (*LeaveSpaceResponse, error) {
	return s.service.LeaveSpace(ctx, &LeaveSpaceRequest{SharedObjectId: sharedObjectID})
}

// DeleteAccount deletes the provider account for a session index, removing its
// sessions and deleting the volume backing store.
func (s *Session) DeleteAccount(ctx context.Context, sessionIdx uint32) (*DeleteAccountResponse, error) {
	return s.service.DeleteAccount(ctx, &DeleteAccountRequest{SessionIdx: sessionIdx})
}

// RenameSpace updates a space display name.
func (s *Session) RenameSpace(ctx context.Context, sharedObjectID, displayName string) (*RenameSpaceResponse, error) {
	return s.service.RenameSpace(ctx, &RenameSpaceRequest{
		SharedObjectId: sharedObjectID,
		DisplayName:    displayName,
	})
}

// WatchResourcesList returns a stream of the full spaces list snapshots.
func (s *Session) WatchResourcesList(ctx context.Context) (SRPCSessionResourceService_WatchResourcesListClient, error) {
	return s.service.WatchResourcesList(ctx, &WatchResourcesListRequest{})
}

// WatchSharedObjectHealth returns a stream of SharedObject health by SharedObject ID.
func (s *Session) WatchSharedObjectHealth(ctx context.Context, sharedObjectID string) (SRPCSessionResourceService_WatchSharedObjectHealthClient, error) {
	return s.service.WatchSharedObjectHealth(ctx, &WatchSharedObjectHealthRequest{SharedObjectId: sharedObjectID})
}

// WatchSyncStatus returns a stream of session sync status snapshots.
func (s *Session) WatchSyncStatus(ctx context.Context) (SRPCSessionResourceService_WatchSyncStatusClient, error) {
	return s.service.WatchSyncStatus(ctx, &WatchSyncStatusRequest{})
}

// WatchStorageStats returns a stream of session storage usage snapshots.
func (s *Session) WatchStorageStats(ctx context.Context) (SRPCSessionResourceService_WatchStorageStatsClient, error) {
	return s.service.WatchStorageStats(ctx, &WatchStorageStatsRequest{})
}

// MountSharedObject mounts a shared object within the session by ID.
// Returns the response containing the resource ID and shared object metadata.
func (s *Session) MountSharedObject(ctx context.Context, sharedObjectID string) (*MountSharedObjectResponse, error) {
	return s.service.MountSharedObject(ctx, &MountSharedObjectRequest{SharedObjectId: sharedObjectID})
}

// WatchLockState returns a stream of the current lock state and updates on changes.
func (s *Session) WatchLockState(ctx context.Context) (SRPCSessionResourceService_WatchLockStateClient, error) {
	return s.service.WatchLockState(ctx, &WatchLockStateRequest{})
}

// SetLockMode changes the session lock mode.
func (s *Session) SetLockMode(ctx context.Context, mode session.SessionLockMode, pin []byte) error {
	_, err := s.service.SetLockMode(ctx, &SetLockModeRequest{Mode: mode, Pin: pin})
	return err
}

// SetDirectP2PEnabled updates the durable Session-local direct transport policy.
func (s *Session) SetDirectP2PEnabled(ctx context.Context, enabled bool) error {
	_, err := s.service.SetDirectP2PEnabled(ctx, &SetDirectP2PEnabledRequest{Enabled: enabled})
	return err
}

// UnlockSession unlocks a PIN-locked session with the given PIN.
func (s *Session) UnlockSession(ctx context.Context, pin []byte) error {
	_, err := s.service.UnlockSession(ctx, &UnlockSessionRequest{Pin: pin})
	return err
}

// LockSession locks a running session, scrubbing the privkey and requiring PIN re-entry.
func (s *Session) LockSession(ctx context.Context) error {
	_, err := s.service.LockSession(ctx, &LockSessionRequest{})
	return err
}

// GetTransferInventory returns the list of spaces on a session for transfer planning.
func (s *Session) GetTransferInventory(ctx context.Context, sessionIdx uint32) (*GetTransferInventoryResponse, error) {
	return s.service.GetTransferInventory(ctx, &GetTransferInventoryRequest{SessionIndex: sessionIdx})
}

// StartTransfer starts a transfer operation between two sessions.
func (s *Session) StartTransfer(ctx context.Context, req *StartTransferRequest) error {
	_, err := s.service.StartTransfer(ctx, req)
	return err
}

// WatchTransferProgress streams transfer state updates for an active transfer.
func (s *Session) WatchTransferProgress(ctx context.Context) (SRPCSessionResourceService_WatchTransferProgressClient, error) {
	return s.service.WatchTransferProgress(ctx, &WatchTransferProgressRequest{})
}

// CancelTransfer stops an in-progress transfer.
func (s *Session) CancelTransfer(ctx context.Context) error {
	_, err := s.service.CancelTransfer(ctx, &CancelTransferRequest{})
	return err
}

// WatchPairingStatus streams pairing state changes during device linking.
func (s *Session) WatchPairingStatus(ctx context.Context) (SRPCSessionResourceService_WatchPairingStatusClient, error) {
	return s.service.WatchPairingStatus(ctx, &WatchPairingStatusRequest{})
}

// CreateLocalPairingOffer generates a WebRTC SDP offer for no-cloud pairing.
func (s *Session) CreateLocalPairingOffer(ctx context.Context) (*CreateLocalPairingOfferResponse, error) {
	return s.service.CreateLocalPairingOffer(ctx, &CreateLocalPairingOfferRequest{})
}

// AcceptLocalPairingOffer accepts a remote offer and returns an answer.
func (s *Session) AcceptLocalPairingOffer(ctx context.Context, offerPayload string) (*AcceptLocalPairingOfferResponse, error) {
	return s.service.AcceptLocalPairingOffer(ctx, &AcceptLocalPairingOfferRequest{OfferPayload: offerPayload})
}

// AcceptLocalPairingAnswer accepts a remote answer to complete the connection.
func (s *Session) AcceptLocalPairingAnswer(ctx context.Context, answerPayload string) (*AcceptLocalPairingAnswerResponse, error) {
	return s.service.AcceptLocalPairingAnswer(ctx, &AcceptLocalPairingAnswerRequest{AnswerPayload: answerPayload})
}

// CreateSpaceInvite creates an invite for a space shared object.
func (s *Session) CreateSpaceInvite(ctx context.Context, req *CreateSpaceInviteRequest) (*CreateSpaceInviteResponse, error) {
	return s.service.CreateSpaceInvite(ctx, req)
}

// RevokeSpaceInvite revokes an invite on a space shared object.
func (s *Session) RevokeSpaceInvite(ctx context.Context, spaceID string, inviteID string) error {
	_, err := s.service.RevokeSpaceInvite(ctx, &RevokeSpaceInviteRequest{SpaceId: spaceID, InviteId: inviteID})
	return err
}

// JoinSpaceViaInvite joins a space using an out-of-band invite message.
func (s *Session) JoinSpaceViaInvite(ctx context.Context, req *JoinSpaceViaInviteRequest) (*JoinSpaceViaInviteResponse, error) {
	return s.service.JoinSpaceViaInvite(ctx, req)
}

// GeneratePairingCode generates a short pairing code for cloud-relay device linking.
func (s *Session) GeneratePairingCode(ctx context.Context) (*GeneratePairingCodeResponse, error) {
	return s.service.GeneratePairingCode(ctx, &GeneratePairingCodeRequest{})
}

// CompletePairing resolves a pairing code from the other device to link.
func (s *Session) CompletePairing(ctx context.Context, code string) (*CompletePairingResponse, error) {
	return s.service.CompletePairing(ctx, &CompletePairingRequest{Code: code})
}

// GetSASEmoji derives the 6-emoji SAS verification sequence for a remote peer.
func (s *Session) GetSASEmoji(ctx context.Context, remotePeerID string) (*GetSASEmojiResponse, error) {
	return s.service.GetSASEmoji(ctx, &GetSASEmojiRequest{RemotePeerId: remotePeerID})
}

// ConfirmSASMatch sends the user's SAS verification decision to the remote peer.
func (s *Session) ConfirmSASMatch(ctx context.Context, confirmed bool) error {
	_, err := s.service.ConfirmSASMatch(ctx, &ConfirmSASMatchRequest{Confirmed: confirmed})
	return err
}

// ConfirmPairing finalizes a verified pairing and persists the remote peer.
func (s *Session) ConfirmPairing(ctx context.Context, remotePeerID string, displayName string) error {
	_, err := s.service.ConfirmPairing(ctx, &ConfirmPairingRequest{
		RemotePeerId: remotePeerID,
		DisplayName:  displayName,
	})
	return err
}

File: sdk/session/session.ts
import { ClientResourceRef } from '@aptre/bldr-sdk/resource/client.js'
import { Resource } from '@aptre/bldr-sdk/resource/resource.js'
import {
  SessionResourceService,
  SessionResourceServiceClient,
} from './session_srpc.pb.js'
import {
  AcceptLocalPairingAnswerResponse,
  AcceptLocalPairingOfferResponse,
  AccessSessionStateAtomRequest,
  CreateLocalPairingOfferResponse,
  CreateSpaceInviteResponse,
  CreateSpaceRequest,
  CreateSpaceResponse,
  GetSessionInfoResponse,
  GetTransferInventoryResponse,
  GetTransferStatusResponse,
  JoinSpaceViaInviteResponse,
  ListSpaceInvitesResponse,
  ListSpaceParticipantsResponse,
  RemoveSpaceParticipantsResponse,
  RenameSpaceRequest,
  MountSharedObjectRequest,
  RevokeSpaceInviteResponse,
  StartTransferRequest,
  WatchLockStateRequest,
  WatchLockStateResponse,
  WatchPairedDevicesResponse,
  WatchPairingStatusResponse,
  WatchResourcesListRequest,
  WatchResourcesListResponse,
  WatchSharedObjectHealthRequest,
  WatchSharedObjectHealthResponse,
  WatchStorageStatsRequest,
  WatchStorageStatsResponse,
  WatchSyncStatusRequest,
  WatchSyncStatusResponse,
  WatchSessionStateAtomsRequest,
  WatchSessionStateAtomsResponse,
  WatchTransferProgressResponse,
} from './session.pb.js'
import { SessionLockMode, SessionRef } from '../../core/session/session.pb.js'
import type {
  SOInviteMessage,
  SOParticipantRole,
} from '../../core/sobject/sobject.pb.js'
import { SharedObject } from '../sobject/sobject.js'
import { SystemStatus } from '../status/status.js'
import { LocalSession } from './local-session.js'
import { SpacewaveSession } from './spacewave-session.js'
import { StateAtom } from '@aptre/bldr-sdk/state/state.js'

// Session is a session resource that provides access to session functionality.
//
// The MountSession directive will remain active until this resource is released.
export class Session extends Resource {
  // sessionRef is the provider identity retained for this mounted resource.
  public readonly sessionRef?: SessionRef

  // service is the session resource service.
  private service: SessionResourceService
  // _spacewave is the lazy-initialized SpacewaveSession wrapper.
  private _spacewave: SpacewaveSession | null = null
  // _localProvider is the lazy-initialized LocalSession wrapper.
  private _localProvider: LocalSession | null = null
  // _systemStatus is the lazy-initialized SystemStatus wrapper.
  private _systemStatus: SystemStatus | null = null

  constructor(resourceRef: ClientResourceRef, sessionRef?: SessionRef) {
    super(resourceRef)
    this.sessionRef = SessionRef.clone(sessionRef) ?? undefined
    this.service = new SessionResourceServiceClient(resourceRef.client)
  }

  // spacewave returns the SpacewaveSession wrapper for session-scoped spacewave RPCs.
  // Both SessionResourceService and SpacewaveSessionResourceService are on the
  // same mux, so the same resourceRef works for both.
  public get spacewave(): SpacewaveSession {
    if (!this._spacewave) {
      this._spacewave = new SpacewaveSession(this.resourceRef)
    }
    return this._spacewave
  }

  // systemStatus returns the SystemStatus wrapper for system status Watch RPCs.
  // SystemStatusService is on the same session mux, so the same resourceRef works.
  public get systemStatus(): SystemStatus {
    if (!this._systemStatus) {
      this._systemStatus = new SystemStatus(this.resourceRef)
    }
    return this._systemStatus
  }

  // getSessionInfo returns information about this session.
  public async getSessionInfo(
    abortSignal?: AbortSignal,
  ): Promise<GetSessionInfoResponse> {
    return await this.service.GetSessionInfo({}, abortSignal)
  }

  // createSpace creates a new Space as a SharedObject within the Session.
  public async createSpace(
    req: CreateSpaceRequest,
    abortSignal?: AbortSignal,
  ): Promise<CreateSpaceResponse> {
    return await this.service.CreateSpace(req, abortSignal)
  }

  // watchResourcesList returns a stream of the full spaces list snapshots.
  public watchResourcesList(
    req?: WatchResourcesListRequest,
    abortSignal?: AbortSignal,
  ): AsyncIterable<WatchResourcesListResponse> {
    return this.service.WatchResourcesList(req ?? {}, abortSignal)
  }

  // watchSharedObjectHealth streams SharedObject health by SharedObject ID.
  public watchSharedObjectHealth(
    req: WatchSharedObjectHealthRequest,
    abortSignal?: AbortSignal,
  ): AsyncIterable<WatchSharedObjectHealthResponse> {
    return this.service.WatchSharedObjectHealth(req, abortSignal)
  }

  // watchSyncStatus streams session sync status snapshots.
  public watchSyncStatus(
    req?: WatchSyncStatusRequest,
    abortSignal?: AbortSignal,
  ): AsyncIterable<WatchSyncStatusResponse> {
    return this.service.WatchSyncStatus(req ?? {}, abortSignal)
  }

  // watchStorageStats streams session storage usage snapshots.
  public watchStorageStats(
    req?: WatchStorageStatsRequest,
    abortSignal?: AbortSignal,
  ): AsyncIterable<WatchStorageStatsResponse> {
    return this.service.WatchStorageStats(req ?? {}, abortSignal)
  }

  // mountSharedObject mounts a shared object and returns the SharedObject resource.
  public async mountSharedObject(
    req: MountSharedObjectRequest,
    abortSignal?: AbortSignal,
  ): Promise<SharedObject> {
    const resp = await this.service.MountSharedObject(req, abortSignal)
    const { resourceId, ...meta } = resp
    return this.resourceRef.createResource(resourceId ?? 0, SharedObject, meta)
  }

  // watchLockState streams the current lock state and updates on changes.
  public watchLockState(
    req?: WatchLockStateRequest,
    abortSignal?: AbortSignal,
  ): AsyncIterable<WatchLockStateResponse> {
    return this.service.WatchLockState(req ?? {}, abortSignal)
  }

  // unlockSession unlocks a PIN-locked session with the given PIN.
  public async unlockSession(
    pin: Uint8Array,
    abortSignal?: AbortSignal,
  ): Promise<void> {
    await this.service.UnlockSession({ pin }, abortSignal)
  }

  // setLockMode changes the session lock mode.
  public async setLockMode(
    mode: SessionLockMode,
    pin?: Uint8Array,
    abortSignal?: AbortSignal,
  ): Promise<void> {
    await this.service.SetLockMode({ mode, pin }, abortSignal)
  }

  // setDirectP2PEnabled updates the durable Session-local direct transport policy.
  public async setDirectP2PEnabled(
    enabled: boolean,
    abortSignal?: AbortSignal,
  ): Promise<void> {
    await this.service.SetDirectP2PEnabled({ enabled }, abortSignal)
  }

  // lockSession locks a running session, scrubbing the privkey and
  // requiring PIN re-entry. Only works when PIN mode is configured.
  public async lockSession(abortSignal?: AbortSignal): Promise<void> {
    await this.service.LockSession({}, abortSignal)
  }

  // generatePairingCode creates a pairing code for P2P device linking.
  public async generatePairingCode(abortSignal?: AbortSignal): Promise<string> {
    const resp = await this.service.GeneratePairingCode({}, abortSignal)
    return resp.code ?? ''
  }

  // completePairing resolves a pairing code to link a remote session.
  public async completePairing(
    code: string,
    abortSignal?: AbortSignal,
  ): Promise<string> {
    const resp = await this.service.CompletePairing({ code }, abortSignal)
    return resp.remotePeerId ?? ''
  }

  // getSASEmoji derives the SAS emoji verification sequence for a remote peer.
  public async getSASEmoji(
    remotePeerId: string,
    abortSignal?: AbortSignal,
  ): Promise<string[]> {
    const resp = await this.service.GetSASEmoji({ remotePeerId }, abortSignal)
    return resp.emoji ?? []
  }

  // confirmSASMatch sends the user's SAS emoji verification decision to the
  // bilateral confirmation exchange running over the bifrost link.
  public async confirmSASMatch(
    confirmed: boolean,
    abortSignal?: AbortSignal,
  ): Promise<void> {
    await this.service.ConfirmSASMatch({ confirmed }, abortSignal)
  }

  // confirmPairing confirms a verified pairing, adding the remote peer as
  // OWNER on all SharedObjects and persisting the paired device.
  public async confirmPairing(
    remotePeerId: string,
    displayName?: string,
    abortSignal?: AbortSignal,
  ): Promise<void> {
    await this.service.ConfirmPairing(
      { remotePeerId, displayName: displayName ?? '' },
      abortSignal,
    )
  }

  // unlinkDevice removes a paired device and revokes its SO access.
  public async unlinkDevice(
    peerId: string,
    abortSignal?: AbortSignal,
  ): Promise<void> {
    await this.service.UnlinkDevice({ peerId }, abortSignal)
  }

  // deleteSpace deletes a space and its associated storage.
  public async deleteSpace(
    sharedObjectId: string,
    abortSignal?: AbortSignal,
  ): Promise<void> {
    await this.service.DeleteSpace({ sharedObjectId }, abortSignal)
  }

  // renameSpace updates the display name metadata for a space.
  public async renameSpace(
    request: RenameSpaceRequest,
    abortSignal?: AbortSignal,
  ): Promise<void> {
    await this.service.RenameSpace(request, abortSignal)
  }

  // deleteAccount deletes the account associated with this session.
  // Cleans ObjectStore keys, removes GC edges, runs volume GC, and
  // removes the session from the list.
  public async deleteAccount(
    sessionIdx: number,
    abortSignal?: AbortSignal,
  ): Promise<void> {
    await this.service.DeleteAccount({ sessionIdx }, abortSignal)
  }

  // accessStateAtom accesses a session-scoped state atom resource.
  // State is persisted in the session's provider account volume.
  public async accessStateAtom(
    request?: AccessSessionStateAtomRequest,
    abortSignal?: AbortSignal,
  ): Promise<StateAtom> {
    const resp = await this.service.AccessStateAtom(request ?? {}, abortSignal)
    return this.resourceRef.createResource(resp.resourceId ?? 0, StateAtom)
  }

  // watchStateAtoms streams the known session state atom store ids on change.
  public watchStateAtoms(
    request?: WatchSessionStateAtomsRequest,
    abortSignal?: AbortSignal,
  ): AsyncIterable<WatchSessionStateAtomsResponse> {
    return this.service.WatchStateAtoms(request ?? {}, abortSignal)
  }

  // localProvider returns a LocalSession wrapper for local-provider-specific RPCs.
  public get localProvider(): LocalSession {
    if (!this._localProvider) {
      this._localProvider = new LocalSession(this.resourceRef)
    }
    return this._localProvider
  }

  // getTransferInventory returns the list of spaces on a session for transfer planning.
  public async getTransferInventory(
    sessionIndex: number,
    abortSignal?: AbortSignal,
  ): Promise<GetTransferInventoryResponse> {
    return await this.service.GetTransferInventory(
      { sessionIndex },
      abortSignal,
    )
  }

  // startTransfer starts a transfer operation between two sessions.
  public async startTransfer(
    req: StartTransferRequest,
    abortSignal?: AbortSignal,
  ): Promise<void> {
    await this.service.StartTransfer(req, abortSignal)
  }

  // watchTransferProgress streams transfer state updates for an active transfer.
  public watchTransferProgress(
    abortSignal?: AbortSignal,
  ): AsyncIterable<WatchTransferProgressResponse> {
    return this.service.WatchTransferProgress({}, abortSignal)
  }

  // cancelTransfer stops an in-progress transfer.
  public async cancelTransfer(abortSignal?: AbortSignal): Promise<void> {
    await this.service.CancelTransfer({}, abortSignal)
  }

  // getTransferStatus returns whether a transfer is active or a checkpoint exists.
  public async getTransferStatus(
    abortSignal?: AbortSignal,
  ): Promise<GetTransferStatusResponse> {
    return await this.service.GetTransferStatus({}, abortSignal)
  }

  // watchPairedDevices streams the list of paired devices from the account settings SO.
  public watchPairedDevices(
    abortSignal?: AbortSignal,
  ): AsyncIterable<WatchPairedDevicesResponse> {
    return this.service.WatchPairedDevices({}, abortSignal)
  }

  // watchPairingStatus streams pairing state changes during device linking.
  public watchPairingStatus(
    abortSignal?: AbortSignal,
  ): AsyncIterable<WatchPairingStatusResponse> {
    return this.service.WatchPairingStatus({}, abortSignal)
  }

  // createSpaceInvite creates an invite for a space shared object.
  // Returns the full response including the invite message and optional short code.
  public async createSpaceInvite(
    spaceId: string,
    role: SOParticipantRole,
    abortSignal?: AbortSignal,
  ): Promise<CreateSpaceInviteResponse> {
    return await this.service.CreateSpaceInvite({ spaceId, role }, abortSignal)
  }

  // listSpaceInvites lists invites on a space shared object.
  public async listSpaceInvites(
    spaceId: string,
    abortSignal?: AbortSignal,
  ): Promise<ListSpaceInvitesResponse> {
    return await this.service.ListSpaceInvites({ spaceId }, abortSignal)
  }

  // listSpaceParticipants lists participants on a space shared object.
  public async listSpaceParticipants(
    spaceId: string,
    abortSignal?: AbortSignal,
  ): Promise<ListSpaceParticipantsResponse> {
    return await this.service.ListSpaceParticipants({ spaceId }, abortSignal)
  }

  // removeSpaceParticipants removes participants in one configuration change.
  public async removeSpaceParticipants(
    spaceId: string,
    peerIds: string[],
    abortSignal?: AbortSignal,
  ): Promise<RemoveSpaceParticipantsResponse> {
    return await this.service.RemoveSpaceParticipants(
      { spaceId, peerIds },
      abortSignal,
    )
  }

  // revokeSpaceInvite revokes an invite on a space shared object.
  public async revokeSpaceInvite(
    spaceId: string,
    inviteId: string,
    abortSignal?: AbortSignal,
  ): Promise<RevokeSpaceInviteResponse> {
    return await this.service.RevokeSpaceInvite(
      { spaceId, inviteId },
      abortSignal,
    )
  }

  // joinSpaceViaInvite joins a space using an out-of-band invite message.
  public async joinSpaceViaInvite(
    inviteMessage: SOInviteMessage,
    targetedInvitationEnvelope?: Uint8Array,
    abortSignal?: AbortSignal,
  ): Promise<JoinSpaceViaInviteResponse> {
    return await this.service.JoinSpaceViaInvite(
      { inviteMessage, targetedInvitationEnvelope },
      abortSignal,
    )
  }

  // createLocalPairingOffer generates a WebRTC SDP offer for no-cloud pairing.
  public async createLocalPairingOffer(
    abortSignal?: AbortSignal,
  ): Promise<CreateLocalPairingOfferResponse> {
    return await this.service.CreateLocalPairingOffer({}, abortSignal)
  }

  // acceptLocalPairingOffer accepts a remote offer and returns an answer.
  public async acceptLocalPairingOffer(
    offerPayload: string,
    abortSignal?: AbortSignal,
  ): Promise<AcceptLocalPairingOfferResponse> {
    return await this.service.AcceptLocalPairingOffer(
      { offerPayload },
      abortSignal,
    )
  }

  // acceptLocalPairingAnswer accepts a remote answer to complete the connection.
  public async acceptLocalPairingAnswer(
    answerPayload: string,
    abortSignal?: AbortSignal,
  ): Promise<AcceptLocalPairingAnswerResponse> {
    return await this.service.AcceptLocalPairingAnswer(
      { answerPayload },
      abortSignal,
    )
  }
}
